### ISSUE TITLE
Snapshot testing NUnit's external API

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit#2413]
Per positive feedback on https://github.com/nunit/nunit/issues/2410#issuecomment-326976168.

- [ ] Add a test that generates a pseudo-C# `pubapi.txt` file off all public and protected symbols in NUnit Framework except the ones in the Internal namespace and fails if it is not identical to the `pubapi.approved.cs` file which is committed to the repo.
    In order to pass the test, the contents of `pubapi.approved.txt` will have to be replaced with the contents of the newly generated `pubapi.txt`. When a changed `pubapi.approved.txt` file shows up in a PR, we must have a compatibility discussion before approving the PR.
    The file will be generated such that any time there is red in the git diff, it represents either a source- or binary-breaking change.

- [ ] Decide whether to .gitignore `pubapi.txt` or keep it beside `pubapi.approved.txt`. The benefit to committing it also is that we could generate it on build rather than on test run and that way you'd have a shorter feedback loop if you make a public API change because it would show up in your git status.

Assigning to myself since I've put a lot of time into exploring this concept, but if someone else really want to claim this they can feel free to edit the assignment and comment.

**Comment by ChrisMaddock** [comment]
The difficult part of this will be https://github.com/nunit/nunit/issues/1426 - it's not totally clear (to me, at least) - what is considered part of our public API, and what isn't. I don't either the 'public keyword' or 'public keyword except the internal namespace' are sufficient definitions - but I'm not really the person to know what is!

As my other comment, in the next major version, it would be great to make as much properly internal as we want to.

**Comment by CharliePoole** [comment]
Well, #2410 had positive feedback, including mine, but also had negative feedback, which you seem to be ignoring. You have not clarified what you consider to be the public API at all. It is not and has never been the `public` keyword. We made all methods public if we wanted to test them, because:
1. The founders came from languages that did not have levels of visibility.
2. We are a test application and wanted to test our stuff
3. There was no supported way except reflection to test non-public members and even after Microsoft invented a way (InternalsVisibleTo) many of us thought that was an ugly kluge and a much greater violation of good coding practice than using public methods

I'm not saying that can't and shouldn't change. I recognize that for the majority of you younger folks working on NUnit now that older point of view doesn't make sense at all. So change it already! But change it in a careful, orderly way:

1. Document what you consider to be the public API. I listed three APIs supported by the framework. Only one of them is documented. You appear to be talking about the API used by folks writing tests, so start with that one. Starting by coding tests is just going to confuse the issue further.

2. As a first cut, I would suggest you investigate all public methods in the NUnit.Framework namespace. They should all be aimed at end users and it's reasonable to consider them part of the API. Look at everything that's there. Does it seem to be part of our API? Or is it in the wrong place? If it's in the wrong place, consider whether there's a chance people are relying on it anyway.

3. SubNamespaces are intended to be a bit more hidden from users. I don't think most of them need to be treated as part of the API, but go through them.

4. Produce a document of some kind, which we can review, that says what you think we need to hold constant from release to release. IMO, we'd be way ahead if we had that, even without the automation.

**Comment by jnm2** [comment]
> You have not clarified what you consider to be the public API at all. It is not and has never been the public keyword.

We are speaking about two different things. There are the documented and curated APIs, but the fact remains that there is a de-facto public API to every IL library– whether intentional and thoughtful or unintentional and thoughtless– which constitutes potential real-world breaking changes at both source level and binary level. We should think **very carefully** before changing any of these, even if they _don't_ fall into one of the three "intentional" APIs you listed.
Therefore this snapshot test is equally useful in bringing them to your attention no matter what your definition of API is. You can choose to acknowledge and approve the real-world break because it's not in one of the three "intentional" APIs. Or you can even choose to be conservative and not approve the break even though it's not in one of the three.

**Comment by CharliePoole** [comment]
@jnm2 No, I think we're speaking about the same thing. I listed three APIs and only _one_ of them is documented. I think you could bring more benefit to us by helping first _define_ what it is we want to support. I'd contribute to that effort.

We already agreed that the methods in the internal namespace are __not__ part of any public API, even though they are public. That's why we are willing to make them internal rather than public. But that issue is itself blocked by the fact that we have not yet defined the extensibility API and some of those methods are in the internal namespace. 

In my view, taking everything that it's _possible_ to use publicly and treating that as an "API" doesn't get us any farther ahead than we are already and might set us back. Your view that we have to be very careful about changing any method that it's _possible_ for someone to use makes sense in dealing with an assembly that was written with your point of view about visibility in mind. But NUnit was not and experienced users are already aware of that.

Aside from that, I'm afraid that what you are outlining will show up so many false negatives that we'll simply stop paying attention to it. We change public methods pretty much all the time.

**Comment by jnm2** [comment]
I was not aware that our extensibility API was internal. The initial snapshot test should not attempt to cover APIs which are internal (if ever). If we want to catch incompatibilities in classes located by reflection convention, that's possibly better handled by integration testing since we own all the parts. 

In light of this, and the fact that you foresee us changing public types all the time, what we'll need is namespace-level granularity. Can we all agree that anything inside the root `NUnit.Framework` should not have source-level or binary-level breaks in externally visible symbols? Are there other namespaces  which we'd also like to cover this way?

**Comment by CharliePoole** [comment]
@jnm2 To be more exact, I don't _forsee_ us changing public methods. It's just what we are now doing and have done in the past. You advocate that we stop doing this. I agree. My view is that we have to go through things one namespace at a time and make decisions.

It's clear that stuff we change as an implementation detail should not be public. We just have to get there. I agree that `NUnit.Framework` is the logical starting place. That holds all our Asserts, Attributes and Syntax helpers. The compatibility problems that started this discussion off are all in that namespace.

**Comment by lilasquared** [comment]
@jnm2 I'd be interested to see what the pseudo-c# file would be like as this was discussed in #2392.  What would be the value of having it be pseudo-c# vs something like a yaml or json document?

**Comment by jnm2** [comment]
Because that way it's easier to read as a C# diff when we review PRs or when someone looks at the pending git changes to the snapshot file.

**Comment by jnm2** [comment]
I'll post an example here when I get far enough.

**Comment by jnm2** [comment]
@lilasquared This is sort of what it will look like: https://pastebin.com/jJX9THHf
It doesn't really give you a good idea because this is what I still have to add:

- [ ] methods
- [ ] constructors
- [ ] properties
- [ ] events
- [ ] operators
- [ ] special attributes
- [ ] generic constraints
- [ ] base and implemented types

But it's basically the entire public and protected (externally binding) API but with no method bodies.

(Proof of concept underway at https://github.com/jnm2/ApiContractGenerator)

**Comment by jnm2** [comment]
Update: https://pastebin.com/nLSzWPBA

**Comment by jnm2** [comment]
I also noticed that these API classes expose these classes from the Internal namespace:

```
NUnit.Framework.RepeatAttribute.RepeatedTestCommand : NUnit.Framework.Internal.Commands.DelegatingTestCommand
NUnit.Framework.RetryAttribute.RetryCommand : NUnit.Framework.Internal.Commands.DelegatingTestCommand
NUnit.Framework.TestCaseData : NUnit.Framework.Internal.TestCaseParameters
NUnit.Framework.TestFixtureData : NUnit.Framework.Internal.TestFixtureParameters
```

So I'll need a way to follow the API graph into the excluded namespace because breaking changes in those classes are breaking changes in the API.

**Comment by jnm2** [comment]
Added methods! https://pastebin.com/nLSzWPBA
This looks a lot closer to what the finished output will be.

**Comment by jnm2** [comment]
I'm surprised at the speed of this thing.

**Comment by jnm2** [comment]
I'm surprised at the speed of this thing reading nunit.framework.dll and outputting 2532 lines!

I just had to know:

   Method |     Mean |     Error |    StdDev |
--------- |---------:|----------:|----------:|
 Generate | 3.909 ms | 0.0808 ms | 0.1282 ms |

**Comment by jnm2** [comment]
Added properties! https://pastebin.com/nLSzWPBA

**Comment by jnm2** [comment]
NUnit doesn't expose events or operators so I'm testing against mscorlib from .NET Framework 4.7. 20,082 lines!

   Method |     Mean |     Error |    StdDev |
--------- |---------:|----------:|----------:|
 Generate | 59.73 ms | 0.3471 ms | 0.3077 ms |

**Comment by jnm2** [comment]
Btw tremendous appreciation for https://sharplab.io/. 

Back in the day I used to learn about IL by coding up a method in a new C# project, switching to release, building, dragging the DLL to ILSpy, clicking down through the tree list, glancing at the IL, removing the DLL  to it would refresh, switching back to C#, etc.

SharpLab is nuts.

**Comment by jnm2** [comment]
Mscorlib apparently doesn't have any generic type parameter constraints. 😆

**Comment by jnm2** [comment]
Switching back to NUnit.

**Comment by jnm2** [comment]
Added generic constraints! https://pastebin.com/nLSzWPBA

**Comment by jnm2** [comment]
Whew, attributes took several times longer than everything else put together so far. Mostly because it requires you to parse serialized assembly-qualified type names, locate the referenced assembly on disk, and read that assembly too to determine enum field names and values.

Current results: https://pastebin.com/nLSzWPBA

**Comment by lilasquared** [comment]
@jnm2 this is really cool! I wanted to ask you about this pseudo c# file that you are creating here.  In Jest, the snapshot files that are created are technically legitimate javascript files and essentially export strings that represent the data that is being snapshotted for the unit test.  If snapshot unit testing was a part of the NUnit framework do you think the idea would be for it to work the same way, as some sort of pseudo C# file that is saved with test file?

**Comment by jnm2** [comment]
@lilasquared If it's pseudo C#, it can't be included in the compilation the way the legitimate JavaScript is. It would have to be legitimate C#. I see the benefit of keeping the snapshots in files under source control, but not of adding new C# source-generated files to the test project compilation.

In my mind, two ideal alternatives for snapshot testing are:

 - One or more autogenerated text files in source control, locatable by the snapshot testing configuration
 - Attributes in the _existing_ C# test file holding the snapshot as a serialized string, paired with a Roslyn analyzer that provides a code fix to run the test and update the snapshot. (Or variations on this line.)

**Comment by jnm2** [comment]
If the snapshot can be serialized to a non-string domain-specific model, I could see us experimenting with autogenerating a `SnapshotSource` static method which builds the snapshot (think CodeDOM in .designer.cs but more modern) and links it to the test method via a `SnapshowSourceAttribute`.

**Comment by jnm2** [comment]
In adding integration tests, I've discovered I'm completely blocked for the time being. Blegh.

https://github.com/dotnet/corefx/issues/24026#issuecomment-332624412

**Comment by jnm2** [comment]
Specifically what I'm blocked on is that any time an attribute has an enum value and the enum is defined in a referenced assembly, I'll have no option but to to show it as `(AttributeTargets)68` rather than `AttributeTargets.Class | AttributeTargets.Method` and there's also the slight chance that the enum is not exactly 32 bits wide and causes me to read invalid data into that numeric value or truncate it because I don't know the enum's underlying type either.

Also a cosmetic thing; default parameter values will all say `default` rather than `null` because `null` is only valid if I can prove it's a reference type, whereas `default` is always valid.

This massively complicates my integration tests between the two platforms because they compare the entire string output and these things are a bit cross-cutting.

**Comment by ChrisMaddock** [comment]
> an attribute has an enum value and the enum is defined in a referenced assembly

Is this something we do? Out of interest - where? 

Shame this is blocked! 😞

**Comment by jnm2** [comment]
@ChrisMaddock This happens every time NUnit.Framework uses `AttributeUsage(AttributeTargets.Foo)` or `EditorBrowsable(EditorBrowsable.Never)`, etc.

**Comment by jnm2** [comment]
When used in an attribute argument value, `AttributeTargets.Class | AttributeTargets.Method` gets serialized as the string `System.AttributeTargets, mscorlib, Version=4.0.0.0, PublicKeyToken=b03f5f7f11d50a3a`  (or if targeting .NET Standard, `System.AttributeTargets, System.Runtime, Version=4.0.20.0, PublicKeyToken=b03f5f7f11d50a3a`) and the bytes `0x00000044` with no indication of how many bytes to read.

So I need to locate the enum definition using the serialized string and read the underlying type (so I know how many bytes to read) and the field names and values so I can format the enum nicely.

**Comment by ChrisMaddock** [comment]
Ah, sorry - I fundamentally misunderstood. I thought you were referring to NUnit's own attributes. Thanks! 😄

**Comment by jnm2** [comment]
I added a fallback to resolve things out of whatever folder `typeof(object).Assembly.Location` is in and that seems to be okay for now.

**Comment by jnm2** [comment]
Special attributes recognized and translated back to C#.

Current results: https://pastebin.com/nLSzWPBA

Todo list moved to https://github.com/jnm2/ApiContractGenerator/issues.

Next up: MSBuild integration.

**Comment by jnm2** [comment]
🎉 https://github.com/jnm2/ApiContractGenerator/commit/025aa3caa8124fedcbd8939cd0aa95366f8375f1

**Comment by jnm2** [comment]
The only thing keeping us from being able to use this is that we need a way to include classes in the excluded (`Internal`) namespace. Extending, implementing, returning, or accepting internal types from the external API forces the internal type to be part of the external API in every sense that the external API itself is: It's the API we market and tell people to use and breaking changes to those internal-but-externally-referenced classes will break folks who stick to the marketed API just as much.

**Comment by CharliePoole** [comment]
The general principle you state is clear. An example might let me understand why you are stating it. 😄

**Comment by jnm2** [comment]
@CharliePoole I'm stating it because it applies to these guys:

<details>

```cs
namespace NUnit.Framework
{
    class CategoryAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class CombiningStrategyAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);

        public System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestMethod> BuildFrom(NUnit.Framework.Interfaces.IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }

    class CultureAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);
    }
    
    class DefaultFloatingPointToleranceAttribute
    {
        public void ApplyToContext(NUnit.Framework.Internal.TestExecutionContext context);
    }

    class ExplicitAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class IgnoreAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class OrderAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class ParallelizableAttribute
    {
        public void ApplyToContext(NUnit.Framework.Internal.TestExecutionContext context);

        public override void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class PlatformAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class PropertyAttribute
    {
        public virtual void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    class RepeatAttribute
    {
        public NUnit.Framework.Internal.Commands.TestCommand Wrap(NUnit.Framework.Internal.Commands.TestCommand command);

        public class RepeatedTestCommand : NUnit.Framework.Internal.Commands.DelegatingTestCommand
        {
            public RepeatedTestCommand(NUnit.Framework.Internal.Commands.TestCommand innerCommand, int repeatCount);

            public override NUnit.Framework.Internal.TestResult Execute(NUnit.Framework.Internal.TestExecutionContext context);
        }
    }

    class RetryAttribute
    {
        public NUnit.Framework.Internal.Commands.TestCommand Wrap(NUnit.Framework.Internal.Commands.TestCommand command);

        public class RetryCommand : NUnit.Framework.Internal.Commands.DelegatingTestCommand
        {
            public RetryCommand(NUnit.Framework.Internal.Commands.TestCommand innerCommand, int tryCount);

            public override NUnit.Framework.Internal.TestResult Execute(NUnit.Framework.Internal.TestExecutionContext context);
        }
    }

    class SetUpFixtureAttribute
    {
        public System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestSuite> BuildFrom(NUnit.Framework.Interfaces.ITypeInfo typeInfo);
    }

    class SingleThreadedAttribute
    {
        public void ApplyToContext(NUnit.Framework.Internal.TestExecutionContext context);
    }

    class TestAttribute
    {
        public void ApplyToTest(NUnit.Framework.Internal.Test test);

        public NUnit.Framework.Internal.TestMethod BuildFrom(NUnit.Framework.Interfaces.IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }

    class TestCaseAttribute
    {
        public System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestMethod> BuildFrom(NUnit.Framework.Interfaces.IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }

    public class TestCaseData : NUnit.Framework.Internal.TestCaseParameters
    {
    }

    class TestCaseSourceAttribute
    {
        public System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestMethod> BuildFrom(NUnit.Framework.Interfaces.IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }

    class TestContext
    {
        public NUnit.Framework.Internal.Randomizer Random { get; }

        public TestContext(NUnit.Framework.Internal.TestExecutionContext testExecutionContext);

        class ResultAdapter
        {
            public ResultAdapter(NUnit.Framework.Internal.TestResult result);
        }

        class TestAdapter
        {
            public TestAdapter(NUnit.Framework.Internal.Test test);
        }
    }

    class TestFixtureAttribute
    {
        public System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestSuite> BuildFrom(NUnit.Framework.Interfaces.ITypeInfo typeInfo);
    }

    public class TestFixtureData : NUnit.Framework.Internal.TestFixtureParameters
    {
    }

    class TestFixtureSourceAttribute
    {
        public System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestSuite> BuildFrom(NUnit.Framework.Interfaces.ITypeInfo typeInfo);
    }

    interface IApplyToContext
    {
        void ApplyToContext(NUnit.Framework.Internal.TestExecutionContext context);
    }

    interface IApplyToTest
    {
        void ApplyToTest(NUnit.Framework.Internal.Test test);
    }

    interface ICommandWrapper
    {
        NUnit.Framework.Internal.Commands.TestCommand Wrap(NUnit.Framework.Internal.Commands.TestCommand command);
    }

    public interface IFixtureBuilder
    {
        System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestSuite> BuildFrom(ITypeInfo typeInfo);
    }

    interface ISimpleTestBuilder
    {
        NUnit.Framework.Internal.TestMethod BuildFrom(IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }

    interface ISuiteBuilder
    {
        NUnit.Framework.Internal.TestSuite BuildFrom(ITypeInfo typeInfo);
    }

    interface ITestBuilder
    {
        System.Collections.Generic.IEnumerable<NUnit.Framework.Internal.TestMethod> BuildFrom(IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }

    interface ITestCaseBuilder
    {
        NUnit.Framework.Internal.Test BuildFrom(IMethodInfo method, NUnit.Framework.Internal.Test suite);

        bool CanBuildFrom(IMethodInfo method, NUnit.Framework.Internal.Test suite);
    }
}
```

</details>

Right now this tool would mislead us into thinking we aren't making breaking changes when in fact we might be.

**Comment by CharliePoole** [comment]
AH! `Test` is a bit of an odd thing. The abstract class was once part of the API. I introduced `ITest` but continued to use `Test` in the interface - quite probably a mistake.

So... is cleaning up our public / private stuff a prerequisite to using your tool or is your tool something to help us perform that cleanup... or both?

I can see that __some people__ would consider changing this interface to use `ITest` a breaking change. Ditto for making it explicit. I personally would not, but I think my view on what is properly considered breaking is the minority view right now.

A final decision would require a careful review of `Test` but I'd be happy enough to see it become more hidden if we didn't find anything "really breaking" in our review.

**Comment by CharliePoole** [comment]
Note that even if we changed `ApplyTo` to take an `ITest`, the author of an attribute can always cast it to `Test` or even to `TestMethod` or `TestSuite` if they like. Based on past experience, I would expect somebody to come along and call changes to the derived class breaking as well.

**Comment by jnm2** [comment]
> So... is cleaning up our public / private stuff a prerequisite to using your tool or is your tool something to help us perform that cleanup... or both?

I was thinking the former is unrealistic if the tool is to be useful and the latter could come later in the form of opt-in/opt-out warnings, but for right now all I want to implement is recursively pulling in types from ignored namespaces that are referenced in non-ignored namespaces. We should at least *see* them. The philosophy I want with this tool is not to prevent breaking changes but to better understand and review them.

On that note, it would be an actual API-breaking change* to replace `Test` with `ITest` but not automatically one I'd be opposed to if we can communicate to end users clearly enough.

* meaning: will cause real-world code to stop compiling even if they only stick to the marketed API

If someone take it upon themselves to cast an object to an `Internal` type, I believe that's no different than if they instantiate or call other methods on an internal type. Both require wandering in, exploring, and including a new `using NUnit.Framework.Internal` statement.

**Comment by CharliePoole** [comment]
Sounds like we agree. If somebody wants to examine the specific change (Test => ITest) I'd favor it.

**Comment by jnm2** [comment]
I finished the implementation, tests pending. I didn't quite believe the results but everything I check makes sense.

Because our non-internal API leaks these types, as shown [above](https://github.com/nunit/nunit/issues/2413#issuecomment-341281452):

- NUnit.Framework.Internal.Commands.DelegatingTestCommand
- NUnit.Framework.Internal.Commands.TestCommand
- NUnit.Framework.Internal.Randomizer
- NUnit.Framework.Internal.Test
- NUnit.Framework.Internal.TestCaseParameters
- NUnit.Framework.Internal.TestExecutionContext
- NUnit.Framework.Internal.TestFixtureParameters
- NUnit.Framework.Internal.TestMethod
- NUnit.Framework.Internal.TestResult
- NUnit.Framework.Internal.TestSuite

That means that source- or binary-breaking changes in the following API also become source- or binary-breaking changes to people who only use type names from the non-internal namespaces. We have to consider all this also our non-internal API:

<details>

```cs
namespace NUnit.Framework.Internal
{
    public class Randomizer : System.Random
    {
        public const string DefaultStringChars = "abcdefghijkmnopqrstuvwxyzABCDEFGHJKLMNOPQRSTUVWXYZ0123456789_";

        public static int InitialSeed { get; set; }

        public static Randomizer CreateRandomizer();

        public static Randomizer GetRandomizer(System.Reflection.MemberInfo member);

        public static Randomizer GetRandomizer(System.Reflection.ParameterInfo parameter);

        public Randomizer();

        public Randomizer(int seed);

        public string GetString(int outputLength, string allowedChars);

        public string GetString(int outputLength);

        public string GetString();

        public bool NextBool();

        public bool NextBool(double probability);

        public byte NextByte();

        public byte NextByte(byte max);

        public byte NextByte(byte min, byte max);

        public System.Decimal NextDecimal();

        public System.Decimal NextDecimal(System.Decimal max);

        public System.Decimal NextDecimal(System.Decimal min, System.Decimal max);

        public double NextDouble(double max);

        public double NextDouble(double min, double max);

        public object NextEnum(System.Type type);

        public T NextEnum<T>();

        public float NextFloat();

        public float NextFloat(float max);

        public float NextFloat(float min, float max);

        public System.Guid NextGuid();

        public long NextLong();

        public long NextLong(long max);

        public long NextLong(long min, long max);

        [System.CLSCompliant(false)]
        public sbyte NextSByte();

        [System.CLSCompliant(false)]
        public sbyte NextSByte(sbyte max);

        [System.CLSCompliant(false)]
        public sbyte NextSByte(sbyte min, sbyte max);

        public short NextShort();

        public short NextShort(short max);

        public short NextShort(short min, short max);

        [System.CLSCompliant(false)]
        public uint NextUInt();

        [System.CLSCompliant(false)]
        public uint NextUInt(uint max);

        [System.CLSCompliant(false)]
        public uint NextUInt(uint min, uint max);

        [System.CLSCompliant(false)]
        public ulong NextULong();

        [System.CLSCompliant(false)]
        public ulong NextULong(ulong max);

        [System.CLSCompliant(false)]
        public ulong NextULong(ulong min, ulong max);

        [System.CLSCompliant(false)]
        public ushort NextUShort();

        [System.CLSCompliant(false)]
        public ushort NextUShort(ushort max);

        [System.CLSCompliant(false)]
        public ushort NextUShort(ushort min, ushort max);
    }

    public abstract class Test : NUnit.Framework.Interfaces.ITest, NUnit.Framework.Interfaces.IXmlNodeBuilder, System.IComparable
    {
        protected NUnit.Framework.Interfaces.ITypeInfo DeclaringTypeInfo;

        public static string IdPrefix { get; set; }

        public abstract object[] Arguments { get; }

        public string ClassName { get; }

        public virtual object Fixture { get; set; }

        public string FullName { get; set; }

        public abstract bool HasChildren { get; }

        public string Id { get; set; }

        public bool IsSuite { get; }

        public NUnit.Framework.Interfaces.IMethodInfo Method { get; set; }

        public virtual string MethodName { get; }

        public string Name { get; set; }

        public NUnit.Framework.Interfaces.ITest Parent { get; set; }

        public NUnit.Framework.Interfaces.IPropertyBag Properties { get; }

        public NUnit.Framework.Interfaces.RunState RunState { get; set; }

        public int Seed { get; set; }

        public System.Reflection.MethodInfo[] SetUpMethods { get; protected set; }

        public System.Reflection.MethodInfo[] TearDownMethods { get; protected set; }

        public virtual int TestCaseCount { get; }

        public abstract System.Collections.Generic.IList<NUnit.Framework.Interfaces.ITest> Tests { get; }

        public virtual string TestType { get; }

        public NUnit.Framework.Interfaces.ITypeInfo TypeInfo { get; }

        public abstract string XmlElementName { get; }

        protected Test(string name);

        protected Test(string pathName, string name);

        protected Test(NUnit.Framework.Interfaces.ITypeInfo typeInfo);

        protected Test(NUnit.Framework.Interfaces.IMethodInfo method);

        public abstract NUnit.Framework.Interfaces.TNode AddToXml(NUnit.Framework.Interfaces.TNode parentNode, bool recursive);

        public void ApplyAttributesToTest(System.Reflection.ICustomAttributeProvider provider);

        public int CompareTo(object obj);

        public virtual TAttr[] GetCustomAttributes<TAttr>(bool inherit) where TAttr : class;

        public void MakeInvalid(string reason);

        public abstract TestResult MakeTestResult();

        protected void PopulateTestNode(NUnit.Framework.Interfaces.TNode thisNode, bool recursive);

        public NUnit.Framework.Interfaces.TNode ToXml(bool recursive);
    }

    public class TestCaseParameters : TestParameters, NUnit.Framework.Interfaces.ITestCaseData, NUnit.Framework.Interfaces.ITestData, NUnit.Framework.Interfaces.IApplyToTest
    {
        public object ExpectedResult { get; set; }

        public bool HasExpectedResult { get; set; }

        public TestCaseParameters();

        public TestCaseParameters(System.Exception exception);

        public TestCaseParameters(object[] args);

        public TestCaseParameters(NUnit.Framework.Interfaces.ITestCaseData data);
    }

    public class TestExecutionContext : NUnit.Compatibility.LongLivedMarshalByRefObject, System.Runtime.Remoting.Messaging.ILogicalThreadAffinative
    {
        public static TestExecutionContext CurrentContext { get; }

        public System.Globalization.CultureInfo CurrentCulture { get; set; }

        public System.Security.Principal.IPrincipal CurrentPrincipal { get; set; }

        public TestResult CurrentResult { get; set; }

        public Test CurrentTest { get; set; }

        public System.Globalization.CultureInfo CurrentUICulture { get; set; }

        public NUnit.Framework.Constraints.ValueFormatter CurrentValueFormatter { get; }

        public NUnit.Framework.Constraints.Tolerance DefaultFloatingPointTolerance { get; set; }

        public NUnit.Framework.Internal.Execution.IWorkItemDispatcher Dispatcher { get; set; }

        public TestExecutionStatus ExecutionStatus { get; set; }

        public bool IsSingleThreaded { get; set; }

        public System.IO.TextWriter OutWriter { get; }

        public NUnit.Framework.ParallelScope ParallelScope { get; set; }

        public Randomizer RandomGenerator { get; }

        public long StartTicks { get; set; }

        public System.DateTime StartTime { get; set; }

        public bool StopOnError { get; set; }

        public int TestCaseTimeout { get; set; }

        public object TestObject { get; set; }

        public NUnit.Framework.Internal.Execution.TestWorker TestWorker { get; }

        public System.Collections.Generic.List<NUnit.Framework.ITestAction> UpstreamActions { get; }

        public TestExecutionContext();

        public TestExecutionContext(TestExecutionContext other);

        public void AddFormatter(NUnit.Framework.Constraints.ValueFormatterFactory formatterFactory);

        public void EstablishExecutionEnvironment();

        public void IncrementAssertCount();

        public void IncrementAssertCount(int count);

        [System.Security.SecurityCritical]
        public override object InitializeLifetimeService();

        public void UpdateContextFromEnvironment();
    }

    public enum TestExecutionStatus : int
    {
        Running = 0,
        StopRequested = 1,
        AbortRequested = 2
    }

    public class TestFixtureParameters : TestParameters, NUnit.Framework.Interfaces.ITestFixtureData, NUnit.Framework.Interfaces.ITestData
    {
        public System.Type[] TypeArgs { get; }

        public TestFixtureParameters();

        public TestFixtureParameters(System.Exception exception);

        public TestFixtureParameters(params object[] args);

        public TestFixtureParameters(NUnit.Framework.Interfaces.ITestFixtureData data);
    }

    public class TestMethod : Test
    {
        public override object[] Arguments { get; }

        public override bool HasChildren { get; }

        public override string MethodName { get; }

        public override System.Collections.Generic.IList<NUnit.Framework.Interfaces.ITest> Tests { get; }

        public override string XmlElementName { get; }

        public TestMethod(NUnit.Framework.Interfaces.IMethodInfo method);

        public TestMethod(NUnit.Framework.Interfaces.IMethodInfo method, Test parentSuite);

        public override NUnit.Framework.Interfaces.TNode AddToXml(NUnit.Framework.Interfaces.TNode parentNode, bool recursive);

        public override TestResult MakeTestResult();
    }

    public abstract class TestParameters : NUnit.Framework.Interfaces.ITestData, NUnit.Framework.Interfaces.IApplyToTest
    {
        public object[] Arguments { get; }

        public object[] OriginalArguments { get; }

        public NUnit.Framework.Interfaces.IPropertyBag Properties { get; }

        public NUnit.Framework.Interfaces.RunState RunState { get; set; }

        public string TestName { get; set; }

        public TestParameters();

        public TestParameters(object[] args);

        public TestParameters(System.Exception exception);

        public TestParameters(NUnit.Framework.Interfaces.ITestData data);

        public void ApplyToTest(Test test);
    }

    public abstract class TestResult : NUnit.Framework.Interfaces.ITestResult, NUnit.Framework.Interfaces.IXmlNodeBuilder
    {
        protected int InternalAssertCount;

        protected System.Threading.ReaderWriterLockSlim RwLock;

        public int AssertCount { get; }

        public System.Collections.Generic.IList<NUnit.Framework.Interfaces.AssertionResult> AssertionResults { get; }

        public abstract System.Collections.Generic.IEnumerable<NUnit.Framework.Interfaces.ITestResult> Children { get; }

        public double Duration { get; set; }

        public System.DateTime EndTime { get; set; }

        public abstract int FailCount { get; }

        public virtual string FullName { get; }

        public abstract bool HasChildren { get; }

        public abstract int InconclusiveCount { get; }

        public string Message { get; }

        public virtual string Name { get; }

        public string Output { get; }

        public System.IO.TextWriter OutWriter { get; }

        public abstract int PassCount { get; }

        public int PendingFailures { get; }

        public NUnit.Framework.Interfaces.ResultState ResultState { get; }

        public abstract int SkipCount { get; }

        public virtual string StackTrace { get; }

        public System.DateTime StartTime { get; set; }

        public NUnit.Framework.Interfaces.ITest Test { get; }

        public System.Collections.Generic.ICollection<NUnit.Framework.Interfaces.TestAttachment> TestAttachments { get; }

        public abstract int WarningCount { get; }

        public NUnit.Framework.Interfaces.AssertionStatus WorstAssertionStatus { get; }

        public TestResult(NUnit.Framework.Interfaces.ITest test);

        public virtual NUnit.Framework.Interfaces.TNode AddToXml(NUnit.Framework.Interfaces.TNode parentNode, bool recursive);

        public void RecordAssertion(NUnit.Framework.Interfaces.AssertionResult assertion);

        public void RecordAssertion(NUnit.Framework.Interfaces.AssertionStatus status, string message, string stackTrace);

        public void RecordAssertion(NUnit.Framework.Interfaces.AssertionStatus status, string message);

        public void RecordException(System.Exception ex);

        public void RecordException(System.Exception ex, NUnit.Framework.Interfaces.FailureSite site);

        public void RecordTearDownException(System.Exception ex);

        public void RecordTestCompletion();

        public void SetResult(NUnit.Framework.Interfaces.ResultState resultState);

        public void SetResult(NUnit.Framework.Interfaces.ResultState resultState, string message);

        public void SetResult(NUnit.Framework.Interfaces.ResultState resultState, string message, string stackTrace);

        public NUnit.Framework.Interfaces.TNode ToXml(bool recursive);
    }

    public class TestSuite : Test
    {
        public override object[] Arguments { get; }

        public override bool HasChildren { get; }

        protected bool MaintainTestOrder { get; set; }

        public System.Reflection.MethodInfo[] OneTimeSetUpMethods { get; protected set; }

        public System.Reflection.MethodInfo[] OneTimeTearDownMethods { get; protected set; }

        public override int TestCaseCount { get; }

        public override System.Collections.Generic.IList<NUnit.Framework.Interfaces.ITest> Tests { get; }

        public override string XmlElementName { get; }

        public TestSuite(string name);

        public TestSuite(string parentSuiteName, string name);

        public TestSuite(NUnit.Framework.Interfaces.ITypeInfo fixtureType, object[] arguments = null);

        public TestSuite(System.Type fixtureType);

        public TestSuite(TestSuite suite, NUnit.Framework.Interfaces.ITestFilter filter);

        public void Add(Test test);

        public override NUnit.Framework.Interfaces.TNode AddToXml(NUnit.Framework.Interfaces.TNode parentNode, bool recursive);

        protected void CheckSetUpTearDownMethods(System.Reflection.MethodInfo[] methods);

        public override TestResult MakeTestResult();

        public void Sort();
    }
}
namespace NUnit.Framework.Internal.Commands
{
    public abstract class DelegatingTestCommand : TestCommand
    {
        protected TestCommand innerCommand;

        protected DelegatingTestCommand(TestCommand innerCommand);
    }

    public class SetUpTearDownItem
    {
        public bool HasMethods { get; }

        public SetUpTearDownItem(System.Collections.Generic.IList<System.Reflection.MethodInfo> setUpMethods, System.Collections.Generic.IList<System.Reflection.MethodInfo> tearDownMethods);

        public void RunSetUp(NUnit.Framework.Internal.TestExecutionContext context);

        public void RunTearDown(NUnit.Framework.Internal.TestExecutionContext context);
    }

    public abstract class TestCommand
    {
        public NUnit.Framework.Internal.Test Test { get; }

        public TestCommand(NUnit.Framework.Internal.Test test);

        public abstract NUnit.Framework.Internal.TestResult Execute(NUnit.Framework.Internal.TestExecutionContext context);
    }
}
namespace NUnit.Framework.Internal.Execution
{
    public interface IWorkItemDispatcher
    {
        void CancelRun(bool force);

        void Dispatch(WorkItem work);

        void Start(WorkItem topLevelWorkItem);
    }

    public enum ParallelExecutionStrategy : int
    {
        Direct = 0,
        Parallel = 1,
        NonParallel = 2
    }

    public class TestWorker
    {
        public bool IsAlive { get; }

        public string Name { get; }

        public WorkItemQueue WorkQueue { get; }

        public event TestWorker.TestWorkerEventHandler Busy;

        public event TestWorker.TestWorkerEventHandler Idle;

        public TestWorker(WorkItemQueue queue, string name);

        public void Cancel(bool force);

        public void Start();

        public delegate void TestWorkerEventHandler(TestWorker worker, WorkItem work);
    }

    public abstract class WorkItem
    {
        public NUnit.Framework.Internal.TestExecutionContext Context { get; }

        public virtual ParallelExecutionStrategy ExecutionStrategy { get; }

        public NUnit.Framework.Interfaces.ITestFilter Filter { get; }

        public virtual string Name { get; }

        public NUnit.Framework.ParallelScope ParallelScope { get; }

        public NUnit.Framework.Internal.TestResult Result { get; protected set; }

        public WorkItemState State { get; }

        public NUnit.Framework.Internal.Test Test { get; }

        public TestWorker TestWorker { get; }

        public event System.EventHandler Completed;

        public WorkItem(NUnit.Framework.Internal.Test test, NUnit.Framework.Interfaces.ITestFilter filter);

        public WorkItem(WorkItem wrappedItem);

        protected System.Collections.Generic.List<NUnit.Framework.Internal.Commands.SetUpTearDownItem> BuildSetUpTearDownList(System.Reflection.MethodInfo[] setUpMethods, System.Reflection.MethodInfo[] tearDownMethods);

        public virtual void Cancel(bool force);

        protected void ChangeResult(NUnit.Framework.Interfaces.ResultState resultState, string message);

        public virtual void Execute();

        public void InitializeContext(NUnit.Framework.Internal.TestExecutionContext context);

        public void MarkNotRunnable(string reason);

        protected abstract void PerformWork();

        protected void WorkItemComplete();
    }

    public class WorkItemQueue
    {
        public bool IsEmpty { get; }

        public bool IsParallelQueue { get; }

        public int ItemsProcessed { get; }

        public string Name { get; }

        public WorkItemQueueState State { get; }

        public System.Threading.ApartmentState TargetApartment { get; }

        public WorkItemQueue(string name, bool isParallel, System.Threading.ApartmentState apartment);

        public WorkItem Dequeue();

        public void Enqueue(WorkItem work);

        public void Pause();

        public void Start();

        public void Stop();
    }

    public enum WorkItemQueueState : int
    {
        Paused = 0,
        Running = 1,
        Stopped = 2
    }

    public enum WorkItemState : int
    {
        Ready = 0,
        Running = 1,
        Complete = 2
    }
}
```

</details>

**Comment by jnm2** [comment]
Updated. I overlooked that method parameter types do not count when recursively unignoring types. They must be in a return position to be followed. The inverse is true of delegates. That cuts it down nicely.

**Comment by CharliePoole** [comment]
So, what is the action to be taken to resolve this issue?

**Comment by jnm2** [comment]
Here's an example of what it looks like to use the MSBuild package right now: https://github.com/jnm2/nunit/commit/0d6ec2c15220151ee28ac5a9653dd0ec5bf1aa72

This is what it looks like (right now) to add it to the project: https://github.com/jnm2/nunit/commit/ba083191c4db906a2ddf122087ffd251b543b0f6 (After we make the csproj SDK switch, that entire commit will be five lines in the (only) .csproj.)

**Comment by jnm2** [comment]
@CharliePoole Sorry, it's late and I missed your comment.

> So, what is the action to be taken to resolve this issue?

I'm not sure yet. Things like Randomizer and even TestCommand aren't really internal; we've never treated them that way, and yet removing `Internal` from the namespace risks breaking code that fully qualifies the names. We could do the move and leave behind stubs that inherit and have an obsoletion warning, but that's ugly. Or we could continue with the status quo (minus perhaps this easy fix: https://github.com/jnm2/nunit/commit/0d6ec2c15220151ee28ac5a9653dd0ec5bf1aa72)

**Comment by CharliePoole** [comment]
Shouldn't the adding of the room be separate from changes to the API?

**Comment by ChrisMaddock** [comment]
> We could do the move and leave behind stubs that inherit and have an obsoletion warning, but that's ugly. 

I quite like this. This has often been a source of contributor pain - people not knowing what is and isn't intended to be internal - so I'd like us to work towards a point where the internal/public division is clear, and we could make internal things actually internal a theoretical 4.x release.

**Comment by CharliePoole** [comment]
I also like this approach for things we can't just fix immediately. My concern is only that I'm simultaneously hearing about issues that seem to be directly fixable and the energy seems to be going toward figuring out how to postpone fixing them rather than fixing them.

I'd like to see some of our effort go into defining low-hanging fruit that we can deal with. @jnm2 has already identified a number of such things, so let's spin off issues to fix them.

IOW it seems to me we are too quickly saying that changes have to wait for 4.0.

For example, why not introduce an interface and make randomizer internal. Why not stop using Test in many places where an ITest can do the job? 

I'd also like to take this a step further and use the tool with different public starting sets. For example we could use it to separately define the extensibility API. By doing that, we could have separate policies for APIs aimed at different groups of users.

@jnm2 You developed the tool separately and it is for general use. So I assume this issue would eventually lead to a PR that adds it as a target in our build. Are we at the point where you are ready to do that? I also see that your personal nunit does already has some low-hanging visibility changes. Will you be doing a PR for those as well?

**Comment by jnm2** [comment]
Yikes!! I realized this morning that I stayed up past 2 am twice last night without realizing it! That last comment was actually 3 am in terms of what I'm used to... :flushed:

> Shouldn't the adding of the room be separate from changes to the API?

I'm sorry, my brain isn't computing that... what's a room?

I think turning the randomizer into an interface is a great idea because the chance that someone has written `NUnit.Framework.Internal.Randomizer` in their code is extremely low. They'd have to be inheriting from it or passing it around their code.

However, anyone using TestCommand will have written `NUnit.Framework.Internal.TestCommand` or imported the namespace- there's no other way to use that type. They are basically forced into doing so by the non-internal `IWrap*` interfaces. I don't see a way to avoid breaking that, since the parameter and return type must match exactly, preventing the inheriting stub trick from working.

So to answer "why not stop using Test in many places where an ITest can do the job," I guess because in the case of Test, people's code will break with no explanation other than release notes, and stubs with obsoletion messages. And stubs won't even compile in the case of implementing the test builder interfaces.

If we shipped an analyzer in the NUnit package we could make upgrading as easy as <kbd>Ctrl</kbd>+<kbd>.</kbd> > Fix all in project. Obsoletion stubs would be totally unnecessary. That's a little work but also allows a cleaner break.

----

I'm not quite ready for the PR that adds this MSBuild target to NUnit.

 - I need to refactor, test and merge the un-ignoring code and address delegate input parameters since internal types can be inferred without you typing `Internal`.

 - Until the new csproj SDK is in use, we don't have a good way to use this MSBuild target. Build.cake thinks we use packages.config and VS 2015 and 2017 think we use project.json. I'm going to wait until that PR goes through and we have a consistency or this will never work reliably for people.

 - The MSBuild task is only part of the story. We need a test in place that compares the tested framework DLLs to the pubapi.txt files and fails if they've been manually re-edited (or someone undoes the changes in git). Otherwise, it'll escape our notice in the PR and the reversed changes will be regenerated and will show up as pending changes every time we build a branch based on that PR. I haven't decided whether we should do this in Cake or as a test-project test. Opinions?

**Comment by CharliePoole** [comment]
LOL! That's autocorrect for "tool".

**Comment by CharliePoole** [comment]
Randomizer versus TestCommand is also a great illustration of the difference between two kinds of users and two APIs.

Randomizer is used (through TestContext) by nunit users in general, folks writing tests.

TestCommand is only used for extenders of NUnit, who are also likely to be users, of course.

The group of non-extending NUnit users is probably something between 100 and 1000 times the size of the group of extending NUnit users. They are also generally much more experienced users.

Therefore, even though the technical definition of "breaking" - or the various definitions - may be the same for both groups, the policies we use to deal with such breaks can be quite different. 

Changing the IWrap interface will have a big impact on a smaller number of people. I hope we will find the same thing WRT Test and ITest.

I think that a tool protecting us from __unintentionally__ breaking things is a great idea. It's just that I'd rather fix a few things that require breaking stuff a little bit and I don't want to wait!

**Comment by CharliePoole** [comment]
See my last comment on #1426
